File: mahaam-api-go/feat/repo/health_repo.go
package repo

import (
	"errors"
	"mahaam-api/feat/models"
	"mahaam-api/infra/dbs"
)

type HealthRepo interface {
	Create(health *models.Health) int64
	UpdatePulse(id UUID) int64
	UpdateStopped(id UUID) int64
}

type healthRepo struct {
}

func NewHealthRepo() HealthRepo {
	return &healthRepo{}
}

func (r *healthRepo) Create(health *models.Health) int64 {
	query := `
		INSERT INTO x_health (id, api_name, api_version, env_name, node_ip, node_name, started_at)
		VALUES (:id, :api_name, :api_version, :env_name, :node_ip, :node_name, current_timestamp)`
	rows := dbs.Exec(query, health)
	if rows != 1 {
		panic(DBErr{Err: errors.New("Health record not created")})
	}
	return rows
}

func (r *healthRepo) UpdatePulse(id UUID) int64 {
	query := `UPDATE x_health SET pulsed_at = current_timestamp WHERE id = :id`
	return dbs.Exec(query, Param{"id": id})
}

func (r *healthRepo) UpdateStopped(id UUID) int64 {
	query := `UPDATE x_health SET stopped_at = current_timestamp WHERE id = :id`
	return dbs.Exec(query, Param{"id": id})
}

File: mahaam-api-py/infra/monitor/health_repo.py
from typing import Protocol
import uuid
from infra import db
from infra.validation import ProtocolEnforcer
from infra.monitor.monitor_models import Health


class HealthRepo(Protocol):
    def create(self, health: Health) -> int: ...
    def update_pulse(self, id: uuid.UUID) -> int: ...
    def update_stopped(self, id: uuid.UUID) -> int: ...


class DefaultHealthRepo(metaclass=ProtocolEnforcer, protocol=HealthRepo):

    def create(self, health: Health) -> int:
        health_data = {
            "id": str(health.id),
            "api_name": health.api_name,
            "api_version": health.api_version,
            "env_name": health.env_name,
            "node_ip": health.node_ip,
            "node_name": health.node_name
        }

        sql = """
            INSERT INTO x_health (id, api_name, api_version, env_name, node_ip, node_name, started_at) 
            VALUES(:id, :api_name, :api_version, :env_name, :node_ip, :node_name, current_timestamp)"""

        return db.DB.insert(sql, health_data)

    def update_pulse(self, id: uuid.UUID) -> int:
        sql = "UPDATE x_health SET pulsed_at = current_timestamp WHERE id = :id"
        return db.DB.update(sql, {"id": str(id)})

    def update_stopped(self, id: uuid.UUID) -> int:
        sql = "UPDATE x_health SET stopped_at = current_timestamp WHERE id = :id"
        return db.DB.update(sql, {"id": str(id)})
